pkg/storage/upload.go

```diff
@@ -1,34 +1,61 @@
 package storage
 
 import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
 	"net/http"
 	"os"
+	"time"
 )
 
-func UploadFile(url string, file string) error {
+const (
+	chunkSize = 1 << 22 // 4MB in bytes
+)
+
+func UploadFile(url string, filePath string) error {
 	uploadUrl, err := getResumableUploadUrl(url)
 	if err != nil {
 		return err
 	}
 
-	// TODO(john): upload in chunks and with parallelism to increase upload speed.
-	data, err := os.Open(file)
+	file, err := os.Open(filePath)
 	if err != nil {
 		return err
 	}
-	defer data.Close()
-	req, err := http.NewRequest("PUT", uploadUrl, data)
+	defer file.Close()
+
+	// Get the file's size.
+	fileInfo, err := file.Stat()
 	if err != nil {
 		return err
 	}
-	req.Header.Set("Content-Type", "application/zip")
+	fileSize := fileInfo.Size()
 
-	res, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return err
+	log.Println("Uploading file...")
+
+	// Upload file in chunks with retry policy for resilience against transient connectivity errors.
+	for offset := int64(0); offset < fileSize; offset += chunkSize {
+		chunk := make([]byte, chunkSize)
+		n, err := file.Read(chunk)
+		if err != nil && err != io.EOF {
+			return err
```

Comment:
Don't need both log fatal and return err
Cuz log fatal crashes

Reply:
I got rid of the log.Fatal calls (right after pushing the PR) in this file, what do u think about only returning errs here?

Reply:
i prefer to only log.Fatal at the outer layer (/cmd)